src/effects/native/echoeffect.cpp

```diff
@@ -149,90 +147,108 @@ void EchoEffect::processChannel(const ChannelHandle& handle, EchoGroupState* pGr
+    bool delay_time_is_changing = false;
+    if (delay_length_samples != gs.prev_delay_samples) {
+        delay_time_is_changing = true;
+        gs.next_read_offset = delay_length_samples;
     }
 
-    int read_position = gs.write_position;
-
-    // Feedback the delay buffer and then add the new input.
+    // Ramp changes in the send parameter to avoid audible pops
     const CSAMPLE_GAIN send_delta = (send_amount - gs.prev_send) /
             (numSamples / EchoGroupState::kChannelCount);
     const CSAMPLE_GAIN send_start = send_amount + send_delta;
+
     for (unsigned int i = 0; i < numSamples; i += EchoGroupState::kChannelCount) {
         CSAMPLE_GAIN send_ramped = send_start;
         if (send_delta > 0.0) {
             send_ramped += send_delta * i / EchoGroupState::kChannelCount;
         }
-        gs.delay_buf[gs.write_position] *= feedback_amount;
-        gs.delay_buf[gs.write_position + 1] *= feedback_amount;
-        gs.delay_buf[gs.write_position] += pInput[i] * send_ramped;
-        gs.delay_buf[gs.write_position + 1] += pInput[i + 1] * send_ramped;
-        // Actual delays distort and saturate, so clamp the buffer here.
-        gs.delay_buf[gs.write_position] =
-                SampleUtil::clampSample(gs.delay_buf[gs.write_position]);
-        gs.delay_buf[gs.write_position + 1] =
-                SampleUtil::clampSample(gs.delay_buf[gs.write_position + 1]);
-        INCREMENT_RING(gs.write_position, EchoGroupState::kChannelCount, delay_samples);
-    }
 
-    // Pingpong the output.  If the pingpong value is zero, all of the
-    // math below should result in a simple copy of delay buf to pOutput.
-    for (unsigned int i = 0; i < numSamples; i += EchoGroupState::kChannelCount) {
+        // Write the input plus feedback to the delay buffer
+        // Actual delays distort and saturate, so clamp the samples.
+        gs.delay_buffer[gs.write_position_left] = SampleUtil::clampSample(
+                pInput[i] * send_ramped + gs.wet_sample_left * feedback_amount);
+        gs.delay_buffer[gs.write_position_left+1] = SampleUtil::clampSample(
+                pInput[i+1] * send_ramped + gs.wet_sample_right * feedback_amount);
+
+        // Read from behind the buffer position where samples were just written
+        int read_position_left = gs.write_position_left - gs.active_read_offset;
+        if (read_position_left < 0) {
+            read_position_left += gs.delay_buffer.size();
+        }
+
+        gs.wet_sample_left = gs.delay_buffer[read_position_left];
+        gs.wet_sample_right = gs.delay_buffer[read_position_left+1];
+
+        if (delay_time_is_changing) {
+            read_position_left = gs.write_position_left - gs.next_read_offset;
+            if (read_position_left < 0) {
+                read_position_left += gs.delay_buffer.size();
+            }
+
+            gs.wet_sample_left = SampleUtil::clampSample(
+                    (gs.wet_sample_left + gs.delay_buffer[read_position_left]) / 2.0);
```

Comment:
a-Delay does some curious [crossfading between the samples here](https://github.com/Ardour/ardour/blob/master/libs/plugins/a-delay.lv2/a-delay.c#L459) using `1 / sampleRate`, but I don't understand the reasoning behind that.

Reply:
Without fully understanding the code, that might be for the pitch effect, of e real tape delay if you change the delay time. Without it we might here click sounds.

Reply:
There are some minor click sounds when changing the Time parameter quickly but I am not sure of their source. I tried the crossfading with 1/numSamples technique that a-Delay uses but it does not solve the issue. The size of the engine's buffer is an artifact of the software's configuration and shouldn't have any impact on the sound of an effect.

Reply:
Emulating the pitch shifting of an analog delay would be fun, but I don't know how to implement that.

Reply:
Guitarix has a nice sounding tape delay LV2 plugin but it is written in Faust and does some complicated modeling of tape including wow & flutter. I checked a-Delay again and adjusting its Time parameter does not have the popping sound this does. I'm not sure what is different that is causing the popping here.